LanguageFeatures/Inline-classes/static_analysis_inline_class_A12_t03.dart

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion Let DV be an inline class declaration named Inline with type
+/// parameters X1 extends B1, .. Xs extends Bs. Assume that DV declares a final
+/// instance variable with name id and type R.
+///
+/// We say that the declared representation type of Inline is R, and the
+/// instantiated representation type corresponding to Inline<T1,.. Ts> is
+/// [T1/X1, .. Ts/Xs]R.
+/// ...
+/// Let V be an inline type of the form Inline<T1, .. Ts>, and let R be the
+/// corresponding instantiated representation type. If R is non-nullable then V
+/// is a proper subtype of Object, and V is non-nullable. Otherwise, V is a
+/// proper subtype of Object?, and V is potentially nullable.
+///
+/// @description Checks that if an instantiated representation type `R` is
+/// not non-nullable then it is a proper subtype of `Object?`
+/// @author sgrekhov22@gmail.com
+
+// SharedOptions=--enable-experiment=inline-class
+
+inline class V1 {
+  final int? id;
+  V1(this.id);
+}
+
+inline class V2<T1, T2 extends num?> {
+  final num? id;
+  V2(this.id);
+}
+
+test1<X extends V1?>(X x) {
+  Object? o = x;
+  x?.id;
+}
+
+test2<X extends V2<String?, int?>?>(X x) {
+  Object? o = x;
+  x?.id;
+}
+
+main() {
+  test1<V1>(V1(42));
+  test2<V2<String?, int?>>(V2(42));
+}
```

Comment:
We'd want to check that `V1` is a subtype of `Object?`: `Object? x = V1(24); // OK.`.

Also, we'd want to check that it is a proper subtype: `V1 v1 = Object() as Object?; // Error.`.

But the test would be stronger if we had

```dart
inline class V1 {
  final Object? id;
  V1(this.id);
}
```

because this would allow an expression of type `V1` to have the value `Object()` and the value null, but an expression with static type `Object?` should still not be assignable to `V1`: `V1 x = null as Object?; // Error.`.

This means that we're really checking that `V1` is a proper subtype of `Object?`, we're not just checking whether we can assign an expression of the representation type to a variable of type `V1` (it would be a bug if that's what is being checked, which is the reason why the test shouldn't succeed when we assign an expression of type `Object?`).

Reply:
I still don't really understand what `test1` and `test2` are testing. We need to test that when the instantiated representation type is nullable or potentially nullable (phrased as 'not non-nullable') then the inline type is a proper subtype of `Object?`.

`final Object? id;` works fine: This is a nullable representation type. So we need to check that `V1` is a proper subtype of `Object?`, that is

```
Object? o = V1(null); // OK: Confirms 'subtype'.
V1 v = anyExpressionWhatsoever as Object?; // Error: Confirms 'proper'.
```

In `test1` we have `X` whose type is potentially nullable because of the `?` in the bound. I don't think it's possible to learn anything about the nullability of `V1` by means of code whether the type `X` is used, because `X` as potentially nullable no matter what `V1` is.

`test2` has the same property: No code using `X` will help us learn anything about `V2<String?, int?>`, because `V2<String?, int?>?` is nullable anyway, so `X` is potentially nullable no matter what.

Reply:
Sounds good, thanks!